### DISCUSSION TITLE
Add exception methods to CharacterBody2D to ignore specific collision objects during move_and_slide

### DISCUSSION THREAD

**Discussion by hsandt** [godotengine/godot-proposals#9461] [2D]
I need my character (CharacterBody2D) to be able to hit an enemy collider by jumping at it (vertically like Mario, or in any direction like Sonic), and allow the character to temporarily stop colliding with the enemy collider(s) after that.

The exceptions could be set with a separate method `add_exception`, or as an optional parameter to `move_and_slide`.

It can serve different purpose:
- being able to get through the enemy and continue your way after the hit, even without killing the enemy
- avoid detecting multiple hits when the character slows down after hit and stays on the edge of the enemy collider, or if the enemy is compounded of multiple colliders and we don't want to register extra hits on every body part

Currently, RayCast2D (https://docs.godotengine.org/en/stable/classes/class_raycast2d.html) already has an API to exclude certain colliders, the add_exception method (and opposite to remove one/clear all of them) as well as the more low-level PhysicsRayQueryParameters2D.exclude array (https://docs.godotengine.org/en/stable/classes/class_physicsrayqueryparameters2d.html).
Even ShapeCast2D (https://docs.godotengine.org/en/stable/classes/class_shapecast2d.html) has add_exception, and that's the closest to what CharacterBody2D move_and_slide does (without the more complex sliding part).

For now, I'll be disabling collision layers to temporarily disable collisions, but it can get more cumbersome when there are many entities on each layer and we only want to disable collision for certain entities, not all of them.

Do you know if exceptions on CharacterBody2D motion itself would cause performance issues? Or if it's just complicated to implement? Because if Shapecast2D exceptions just work, I may as well use it and manually move and slide my character based on the physics cast result - but it also means that CharacterBody2D move_and_slide may rely on existing implementation to allow exceptions.

**Comment by AThousandShips** [comment]
Does [`add_collision_exception_with`](https://docs.godotengine.org/en/stable/classes/class_physicsbody2d.html#class-physicsbody2d-method-add-collision-exception-with) work?
